bindings/pyroot_experimental/PyROOT/src/RDataFramePyz.cxx

```diff
@@ -0,0 +1,95 @@
+#include "ROOT/RDF/PyROOTHelpers.hxx" // RResultPtrDummy
+
+#include <utility> // std::pair
+#include <sstream> // std::stringstream
+
+////////////////////////////////////////////////////////////////////////////
+/// \brief Make an RDataFrame from a dictionary of numpy arrays
+/// \param[in] pydata Dictionary with numpy arrays
+///
+/// This function takes a dictionary of numpy arrays and creates an RDataFrame
+/// using the keys as column names and the numpy arrays as data.
+PyObject *PyROOT::MakeRDataFrame(PyObject * /*self*/, PyObject * pydata)
+{
+   if (!pydata) {
+      PyErr_SetString(PyExc_RuntimeError, "Object not convertible: Invalid Python object.");
+      return NULL;
+   }
+
+   if (!PyDict_Check(pydata)) {
+      PyErr_SetString(PyExc_RuntimeError, "Object not convertible: Python object is not a dictionary.");
+      return NULL;
+   }
+
+   // Iterate over dictionary, convert numpy arrays to RVecs and put together interpreter code
+   PyObject *key, *value;
+   Py_ssize_t pos = 0;
+   std::string code = "ROOT::Internal::RDF::NewLazyDataFrame(";
+   auto pyvecs = PyDict_New();
+   while (PyDict_Next(pydata, &pos, &key, &value)) {
+      // Get name of key
+      if (!CPyCppyy_PyUnicode_Check(key)) {
+         PyErr_SetString(PyExc_RuntimeError, "Object not convertible: Dictionary key is not convertible to a string.");
+         return NULL;
+      }
+      std::string keystr = CPyCppyy_PyUnicode_AsString(key);
+
+      // Convert value to RVec and attach to dictionary
+      auto pyvec = PyROOT::AsRVec(NULL, value);
+      if (pyvec == NULL) {
+         PyErr_SetString(PyExc_RuntimeError,
+                         ("Object not convertible: Dictionary entry " + keystr + " is not convertible with AsRVec.").c_str());
+         return NULL;
+      }
+      PyDict_SetItem(pyvecs, key, pyvec);
+
+      // Prepare interpreter code
+      std::string vectype = Cppyy::GetScopedFinalName(((CPyCppyy::CPPInstance*)pyvec)->ObjectIsA());
+      std::string ptrtype = "ROOT::Internal::RDF::RResultPtrDummy<" + vectype + ">";
+      std::stringstream ss;
+      ss << ((CPyCppyy::CPPInstance*)pyvec)->GetObject();
+      auto vecaddress = ss.str();
+      code += "std::pair<std::string," + ptrtype +  ">(\"" + keystr
+           + "\"," + ptrtype + "(*reinterpret_cast<" + vectype+ "*>(" + vecaddress + ")))";
+      code += ",";
+   }
+   code.pop_back();
+   code += ");";
+
+   // Create RDataFrame and build Python proxy
+   auto address = (void*) gInterpreter->Calc(code.c_str());
+   auto klass = (Cppyy::TCppType_t)Cppyy::GetScope("ROOT::RDataFrame");
+   auto pyobj = CPyCppyy::BindCppObject(address, klass);
+
+   // Give Python the ownership of the underlying C++ object
+   ((CPyCppyy::CPPInstance*)pyobj)->PythonOwns();
+
+   // Bind pyobject holding adopted memory to the RVec
+   // TODO: Do we need to keep a reference to the data dictionary or is this
+   // already fixed by the reference from the rvec to the numpy arrays?
```

Comment:
`RVec`s do not prolong the lifetime of the data when they act as views, I think `pyobj` should keep a reference to `pydata`

Reply:
The `AsRVec` does already attach a reference of the original numpy array to itself so that the numpy array cannot go out of scope anymore (unless the rvec dies). So keeping the rvecs should be sufficient so that the data stays in scope. But I really need to double check this ;)

Reply:
Just tested, we loose the data dictionary, however, the actual numpy arrays survive. Still, if the RDF main node goes out of scope, the data dies. So the ownership of the data is not passed to the last node such as done in C++. But I don't think we can fix this.

Reply:
> if the RDF main node goes out of scope, the data dies

This is evil: it's different than for any other dataframe variable both in C++ and python, and especially in python it's common to write things like:
```
df = df.Filter("x > 0")
```
The line above would make the event loop crash but only for numpy dataframes and only if the numpy arrays have been garbage collected.

I think either we figure out a way to make ownership work smoothly no matter what users write, or we clearly document in very large letters -- actually even in the factory name -- that the dataframe is a _view_ over the numpy arrays, and the user is responsible to keep them around.

Reply:
Yep, I agree. Did not think of the `df  = df.something()` case ... Well, we could return something like `df, vecs = ROOT.ROOT.RDF.MakeNumpyDataFrame(...)`. However, this is not very smooth.

Reply:
I need to think about this again in very detail. I'm not sure whether we are missing something here.